homeassistant/components/pi_hole/__init__.py

```diff
@@ -136,3 +178,75 @@ async def async_update_data() -> None:
 async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Unload Pi-hole entry."""
     return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+
+async def determine_api_version(
+    hass: HomeAssistant, entry: dict[str, Any]
+) -> Literal[5, 6]:
+    """Determine the API version of the Pi-hole instance without requiring authentication.
+
+    Neither API v5 or v6 provides an endpoint to check the version without authentication.
+    Version 6 provides other enddpoints that do not require authentication, so we can use those to determine the version
+    version 5 returns an empty list in response to unauthenticated requests.
+    Because we are using endpoints that are not designed for this purpose, we should log liberally to help with debugging.
+    """
+    session = async_get_clientsession(hass, entry[CONF_VERIFY_SSL])
+    holeV6 = Hole(
+        host=entry[CONF_HOST],
+        session=session,
+        location=entry[CONF_LOCATION],
+        protocol="https" if entry[CONF_SSL] else "http",
+        password="wrong_password",
+        version=6,
+    )
+    try:
+        await holeV6.authenticate()
+    except HoleError as ex_v6:
+        if str(ex_v6) == "Authentication failed: Invalid password":
```

Comment:
This is kinda fragile. I'm OK with it for now, particularly since the external library is still part of the HA universe, but there should also be an update to [home-assistant-ecosystem/python-hole](https://github.com/home-assistant-ecosystem/python-hole) to raise more fine-grained exceptions to support this better.

Reply:
Added a comment to reflect this.